trl/experimental/gold/gold_trainer.py

```diff
@@ -0,0 +1,1984 @@
+                    student_logits=shifted_student_logits,
+                    teacher_logits=shifted_teacher_logits,
+                    labels=shifted_labels,
+                    beta=self.beta,
+                )
+
+        if self.use_uld_loss:
+            student_input_ids = inputs["input_ids"]
+
+            # Use the *teacher* labels created above, not the student's.
+            teacher_labels_for_loss = teacher_labels if "teacher_labels" in locals() else inputs["labels"]
+            teacher_input_ids_for_loss = teacher_input_ids if "teacher_input_ids" in locals() else inputs["input_ids"]
+
+            # Create properly masked student labels (fixing batch size > 1 issue)
+            student_labels = inputs["labels"].clone()
+            if hasattr(self.processing_class, "pad_token_id") and self.processing_class.pad_token_id is not None:
+                student_labels[student_labels == self.processing_class.pad_token_id] = -100
+
+            # Also mask pad tokens in teacher labels for consistency
+            if (
+                hasattr(self, "teacher_tokenizer")
+                and hasattr(self.teacher_tokenizer, "pad_token_id")
+                and self.teacher_tokenizer.pad_token_id is not None
+            ):
+                teacher_labels[teacher_labels == self.teacher_tokenizer.pad_token_id] = -100
+
+            loss = self.uld_loss_fn(
+                student_logits=outputs_student.logits,
+                teacher_logits=outputs_teacher.logits,
+                student_labels=student_labels,
+                teacher_labels=teacher_labels_for_loss,
+                student_input_ids=student_input_ids,
+                teacher_input_ids=teacher_input_ids_for_loss,
+            )
+
+            # If ULD hybrid mode produced per-step matched/unmatched components, accumulate them for logging.
+            # Use gradient_accumulation_steps to mirror Trainer's windowing behavior.
+            if hasattr(self.uld_loss_fn, "last_matched_loss") and hasattr(self.uld_loss_fn, "last_unmatched_loss"):
+                try:
+                    ga = max(1, int(self.args.gradient_accumulation_steps))
+                except Exception:
+                    ga = 1
+                step_eq = 1.0 / ga
+                # read scalar values for logging
+                matched_val = (
+                    self.uld_loss_fn.last_matched_loss.item()
+                    if self.uld_loss_fn.last_matched_loss is not None
+                    else 0.0
+                )
+                unmatched_val = (
+                    self.uld_loss_fn.last_unmatched_loss.item()
+                    if self.uld_loss_fn.last_unmatched_loss is not None
+                    else 0.0
+                )
+
+                self._matched_sum += matched_val
+                self._unmatched_sum += unmatched_val
+                self._matched_step_eq += step_eq
+                self._unmatched_step_eq += step_eq
+
+        empty_cache()
```

Comment:
out of curiosity, did you try to ablate this? normally we don't have to handle cache emptying so I'm always surprised to see this